src/spatialmath/utils.rs

```diff
@@ -45,42 +48,394 @@ impl EulerAngles {
+                let rot_quat = rot_quat_unit.quaternion();
+                let z_axis_quat = Quaternion::new(0.0, 0.0, 0.0, 1.0);
+                let test_z = (rot_quat * z_axis_quat) * rot_quat.conjugate();
+                let test_z_imag = test_z.imag();
+
+                let normal_3 = new_z_imag.cross(&test_z_imag);
+                let cos_test = normal_1.dot(&normal_3) / (normal_3.norm() * normal_1.norm());
+                match cos_test {
+                    val2 if (1.0 - val2) < (ANGLE_ACCEPTANCE * ANGLE_ACCEPTANCE) => -1.0 * val,
+                    _ => val
+                }
+            },
+            _ => 0.0
+        };
+    }
+    match new_z.k {
+        val if val < 0.0 => -1.0 * new_x.j.atan2(new_x.i),
+        _ => -1.0 * new_x.j.atan2(new_x.i * -1.0)
+    }
+}
+
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
+pub struct OrientationVector {
+    pub o_vector: UnitVector3<f64>,
+    pub theta: f64
+}
+
+impl OrientationVector {
+    pub fn new(o_x: f64, o_y: f64, o_z: f64, theta: f64) -> Self {
+        let o_vector = UnitVector3::new_normalize(Vector3::new(o_x, o_y, o_z));
+        OrientationVector{o_vector, theta}
+    }
+
+    pub fn to_quaternion(&self) -> Quaternion<f64> {
+        let lat = self.o_vector.z.acos();
+        let lon = match self.o_vector.z {
+            val if 1.0 - val > ANGLE_ACCEPTANCE => self.o_vector.y.atan2(self.o_vector.x),
+            _ => 0.0
+        };
+
+        // convert angles as euler angles (lon, lat, theta) to quaternion
+        // using the zyz rotational order
+        let s: [f64;3] = [
+            (lon / 2.0).sin(),
+            (lat / 2.0).sin(),
+            (self.theta / 2.0).sin()
+        ];
+
+        let c: [f64;3] = [
+            (lon / 2.0).cos(),
+            (lat / 2.0).cos(),
+            (self.theta / 2.0).cos()
+        ];
+
+        let real = c[0]*c[1]*c[2] - s[0]*c[1]*s[2];
+        let i = c[0]*s[1]*s[2] - s[0]*s[1]*c[2];
+        let j = c[0]*s[1]*c[2] + s[0]*s[1]*s[2];
+        let k = s[0]*c[1]*c[2] + c[0]*c[1]*s[2];
+
+        Quaternion::new(real, i, j, k)
+    }
+}
+
+impl ApproxEq for OrientationVector {
+    type Margin = F64Margin;
+
+    fn approx_eq<M: Into<Self::Margin>>(self, other: Self, margin: M) -> bool {
+        let margin = margin.into();
+        let vec_diff = self.o_vector.into_inner() - other.o_vector.into_inner();
+        vec_diff.norm_squared().approx_eq(0.0, margin) && self.theta.approx_eq(other.theta, margin)
+    }
+}
+
+impl From<Quaternion<f64>> for OrientationVector {
+    fn from(quat: Quaternion<f64>) -> Self {
+        let x_quat = Quaternion::new(0.0, -1.0, 0.0, 0.0);
+        let z_quat = Quaternion::new(0.0, 0.0, 0.0, 1.0);
+
+        let conj = quat.conjugate();
+        let new_x = (quat * x_quat) * conj;
+        let new_z = (quat * z_quat) * conj;
+
+        let o_vector = UnitVector3::new_normalize(new_z.imag());
+        let theta = orientation_vector_theta_from_rotated_axes(new_x, new_z);
```

Comment:
[q] Maybe this is my Rust being rusty, Why is the theta logic not written out here and bundled intoa different function? Interested in the reasoning behind the choice. 

They are bundled up together in rdk. I'm just wondering if accessing the theta part of creating the orientation vector could be a problem for users using this library to convert. Can they get to this function on it's own, and is it useful purely on it's own.

Reply:
I wrote it this way because it was easier for me to debug, but I can remove the helper and put the logic directly in the function